src/python/pants/engine/fs.py

```diff
@@ -330,6 +331,64 @@ def from_snapshots(cls, ours: Snapshot, theirs: Snapshot) -> SnapshotDiff:
         return cls(*ours._diff(theirs))
 
 
+class PathMetadataKind(Enum):
+    """Enumeration denoting the kind of path entry represented by a `PathMetadata`."""
+
+    FILE = "file"
+    DIRECTORY = "directory"
+    SYMLINK = "symlink"
+
+
+@dataclass(frozen=True)
+class PathMetadata:
+    """Full metadata of a single path in the repository.
+
+    This dataclass represents the full metadata of a path in the filesystem, but not its contents. This API
+    exists to support certain APIs which have a special need to inspect a path's full metadata. Rule
+    code should normally just use the `Digest` and `Snapshot`-related APIs to capture file content
+    from the repository along with the subset of metadata actually supported across all execution environments
+    (e.g., the `is_executable` flag).
+
+    Some of the fields are optional because not all fields may be supported by the underlying system (e.g.,
+    the repository filesystem or an REAPI directopry tree).
+    """
+
+    path: str
+    kind: str  # see `PathMetadataKind` for supported values
```

Comment:
I would love if this field were of type `PathMetadataKind`. Any examples of how to accomplish this from Python easily?

Reply:
Can this be a Rust enum with `#[pyclass]`? https://pyo3.rs/v0.21.2/class.html#pyclass-enums

Or do dataclasses not play nicely with those?

Or, maybe you can implement `FromPyObject` on a parallel Rust enum?

Reply:
Yes, done.